crates/holochain/src/core/workflow/countersigning_workflow.rs

```diff
@@ -1,332 +1,616 @@
+    pub outcomes: Vec<SessionResolutionOutcome>,
+    /// If we've tried to complete the session using signatures found with agent activity authorities
+    /// and still not succeeded, then the number of attempts is recorded here.
+    pub completion_attempts: usize,
+}
+
+impl Default for SessionResolutionSummary {
+    fn default() -> Self {
+        Self {
+            attempts: 1,
+            last_attempt_at: Timestamp::now(),
+            outcomes: Vec::with_capacity(0),
+            completion_attempts: 0,
+        }
     }
-    Ok(())
 }
 
-/// Countersigning workflow that checks for complete sessions and
-/// pushes the complete ops to validation then messages the signers.
+/// The outcome for a single agent who participated in a countersigning session.
+///
+/// [NUM_AUTHORITIES_TO_QUERY] authorities are made to agent activity authorities for each agent,
+/// and the decisions are collected into [SessionResolutionOutcome::decisions].
+#[derive(Debug, Clone)]
+struct SessionResolutionOutcome {
+    /// The agent who participated in the countersigning session and is the subject of this
+    /// resolution outcome.
+    // Unused until the next PR
+    #[allow(dead_code)]
+    pub agent: AgentPubKey,
+    /// The resolved decision for each authority for the subject agent.
+    // Unused until the next PR
+    #[allow(dead_code)]
+    pub decisions: Vec<SessionCompletionDecision>,
+}
+
+const NUM_AUTHORITIES_TO_QUERY: usize = 3;
+
+#[derive(Clone, Debug, PartialEq)]
+enum SessionCompletionDecision {
+    Complete(Box<SignedActionHashed>),
+    Abandoned,
+    Indeterminate,
+}
+
+#[cfg_attr(feature = "instrument", tracing::instrument(skip_all))]
+#[allow(clippy::too_many_arguments)]
 pub(crate) async fn countersigning_workflow(
     space: Space,
-    network: impl HolochainP2pDnaT,
-    sys_validation_trigger: TriggerSender,
+    network: Arc<impl HolochainP2pDnaT>,
+    cell_id: CellId,
+    signal_tx: Sender<Signal>,
+    self_trigger: TriggerSender,
+    integration_trigger: TriggerSender,
+    publish_trigger: TriggerSender,
 ) -> WorkflowResult<WorkComplete> {
-    // Get any complete sessions.
-    let complete_sessions = space.countersigning_workspace.get_complete_sessions();
-    let mut notify_agents = Vec::with_capacity(complete_sessions.len());
-
-    // For each complete session send the ops to validation.
-    for (agents, ops, actions) in complete_sessions {
-        let non_enzymatic_ops: Vec<_> = ops
-            .into_iter()
-            .filter(|(_hash, dht_op)| dht_op.enzymatic_countersigning_enzyme().is_none())
-            .collect();
-        if !non_enzymatic_ops.is_empty() {
-            incoming_dht_ops_workflow(
+    tracing::debug!(
+        "Starting countersigning workflow, with {} sessions",
+        space
+            .countersigning_workspace
+            .inner
+            .share_ref(|inner| Ok(inner.sessions.len()))
+            .unwrap()
```

Comment:
Because the share type defined in Kitsune and used across Holochain isn't cleanly generic. Even if your inner closure is infallible, you are forced to return a result. It's grim!